usim/_basics/resource.py

```diff
@@ -0,0 +1,208 @@
+    resources:
+
+    .. code:: python3
+
+        # create a limited supply of resources
+        resources = Capacities(cores=8, memory=16000)
+
+        # temporarily remove resources
+        async with resources.borrow(cores=2, money=4000):
+            await computation
+
+    A :py:class:`~.Capacities` guarantees that its resources are conserved and
+    cannot be leaked. Once resources are :py:meth:`~.borrow`\ ed, they can
+    always be returned promptly.
+    """
+    def __init__(self, __zero__: Optional[T] = None, **capacity: T):
+        resources = Resources(__zero__, **capacity)
+        super().__init__(resources, resources.levels)
+        self._available = Tracked(resources.levels)
+
+
+class Resources(BaseResources[T]):
+    r"""
+    Supply of named resources which can be temporarily borrowed or produced/consumed
+
+    The resources and their initial levels are defined
+    when the resource supply is created.
+    Afterwards, the level of resources can be permanently :py:meth:`~.increase`\ d or
+    :py:meth:`~.decrease`\ d as well as temporarily decreased by :py:meth:`borrow`\ ing:
+
+    .. code:: python3
+
+        # create an open supply of resources
+        resources = Resources(cores=8, memory=4000)
+
+        # increase the resource supply available
+        resources.increase(memory=2000)
+
+        # temporarily remove resources
+        async with resources.borrow(cores=2, memory=6000):
+            await computation
+
+        # decrease the resource supply available
+        resources.decrease(cores=4)
+
+    A :py:class:`~.Capacities` guarantees that is always possible
+    to increase the level of available resources.
+    Once resources are :py:meth:`~.borrow`\ ed, they can
+    always be returned promptly.
+    """
+    def __init__(self, __zero__: Optional[T] = None, **capacity: T):
+        if not capacity:  # Note: this should be a type-error not assert for consistency
+            raise TypeError(
+                '%s requires at least 1 keyword-only argument' % self.__class__.__name__
+            )
+        __zero__ = __zero__ if __zero__ is not None else\
+            type(next(iter(capacity.values())))()  # bare type invocation must be zero
+        self._levels_type = __specialise__(__zero__, capacity.keys())
+        self._available = Tracked(self._levels_type(**capacity))
+        assert self._available > self._levels_type.zero,\
+            'initial capacities must be greater than zero'
+
+    async def set(self, **amounts: T):
+        """
+        Set the level of resources
+
+        :param amounts: resource levels to set
+
+        Only levels of resources that are already part of these
+        :py:class:`~.Resources` can be set. Levels cannot be set
+        below zero. If a resource is not specified, its level remains
+        unchanged.
+        """
+        assert self._levels_type.zero <= self._levels_type(**amounts),\
```

Comment:
Here I would assert for `<`. The use case might also include setting a `0` value.

```suggestion
        assert self._levels_type.zero < self._levels_type(**amounts),\
```